### ISSUE TITLE
[lxd] `Bad Request` when trying to launch

### ISSUE THREAD

**Issue by Saviq** [canonical/multipass#1775]
I can't reproduce this any more, but I was in a situation where launching instances failed without a good error message:
```
launch failed: unix://multipass/var/snap/lxd/common/lxd/unix.socket@1.0/virtual-machines?project=multipass: Bad Request
```

This happened after an image downloaded, but before the instance was created (as can be seen in the monitor log).

I _think_ the `Bad Request` is because we're trying to delete a non-existent instance - we should probably catch that. Why wouldn't it launch in the first place, though :shrug:.

[lxc-monitor-edge.log](https://github.com/canonical/multipass/files/5310183/lxc-monitor-edge.log)
[multipass-launch-edge.txt](https://github.com/canonical/multipass/files/5310184/multipass-launch-edge.txt)

**Comment by townsend2010** [comment]
Hmm, the logs aren't very revealing, but it looks like you started to launch an instance and in the middle of the downloading of the image by LXD, you canceled that and then tried to launch another instance and let the download complete this time and then got the `Bad Request`.  Is that accurate?

Also, as an aside but a gripe I have, `lxc monitor` really should have more info when it gets the `message: 'Database error: &errors.errorString{s:"No such object"}'` error as that is really the problem we are trying to debug.

Anyways, let me know if that is an accurate scenario and I'll keep looking.

**Comment by Saviq** [comment]
> Also, as an aside but a gripe I have, `lxc monitor` really should have more info when it gets the `message: 'Database error: &errors.errorString{s:"No such object"}'` error as that is really the problem we are trying to debug.

I could probably get that from daemon with `--debug`.

> Anyways, let me know if that is an accurate scenario and I'll keep looking.

Not sure about the canceling, but that could well be? And yes, subsequently the image download completed, and `Bad Request` was raised

**Comment by townsend2010** [comment]
Ok, I'll inspect the code and see if anything jumps out and also try to reproduce.

**Comment by townsend2010** [comment]
I kind of understand what's going on here, but I'm afraid there isn't much we can do about it.  I don't know what the exact cause of this is, but it doesn't look like LXD will give us much to go by either.

So what's going on is if LXD detects any issue when `POST`ing to create an instance, it will send a `Bad Request` HTTP response with a corresponding JSON reply with details.  But in this case, the details are dearth of any information.

For example, I short-circuited the `LXDVMImageVault` to not download the image if it's missing from the LXD image store and I indeed get the same `Bad Request` at launch.  I then modified `lxd_request()` to print out the JSON response, and this is what LXD gives us:
```
[trace] [lxd request] Got reply: {
    "error": "No such object",
    "error_code": 400,
    "type": "error"
}
```
:facepalm: 

So the only thing we could change here is switch from printing `Bad Request` to `No such object`.  Still not very helpful.  I think we need to follow up with the LXD folks about having better error messages when there are bad requests.
